rust/main/submitter/src/chain_tx_adapter/chains/sealevel/adapter.rs

```diff
@@ -316,12 +323,39 @@ impl AdaptsChain for SealevelTxAdapter {
         Ok(success)
     }
 
+    async fn estimate_tx(&self, tx: &mut Transaction) -> Result<(), SubmitterError> {
+        info!(?tx, "estimating transaction");
+        let not_estimated = tx.precursor();
+        let estimated = self.estimate(not_estimated).await?;
+
+        // If cache does not contain estimate type, insert Simulation type so that it can be used on the first submission
+        {
+            let mut guard = self.estimate_type_cache.lock().await;
+            if guard.get(&tx.id).is_none() {
+                guard.insert(tx.id.clone(), EstimateType::Simulation);
+            }
+        };
+
+        tx.vm_specific_data = VmSpecificTxData::Svm(estimated);
+        info!(?tx, "estimated transaction");
+        Ok(())
+    }
+
     async fn submit(&self, tx: &mut Transaction) -> Result<(), SubmitterError> {
         info!(?tx, "submitting transaction");
-        let not_estimated = tx.precursor();
-        // TODO: the `estimate` call shouldn't happen here - the `Transaction` argument should already contain the precursor,
-        // set in the `build_transactions` method
-        let estimated = self.estimate(not_estimated.clone()).await?;
+
+        let previous = tx.precursor();
+        let estimated = {
+            let mut guard = self.estimate_type_cache.lock().await;
+            match guard.get(&tx.id) {
+                Some(EstimateType::Estimation) | None => self.estimate(previous).await?,
```

Comment:
iiuc, `EstimateType` represents whether cache should be refreshed or not. If so, I'd be inclined to call it slightly differently:
```rust
#[derive(Default)]
enum EstimationCacheFreshness {
    #[default]
    Stale,
    Fresh
}
```

with the default declared, we can then slightly simplify the `match` by doing `match guard.get(&tx.id).unwrap_or_default()`, since we're really matching over two cases (stale or fresh), but having a third case confused me (Estimation, Simulation, None)

Reply:
It makes sense. I'll adopt it.